core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala

```diff
@@ -52,7 +52,7 @@ class MetadataRequestTest extends BaseRequestTest {
     assertNotEquals("Controller id should switch to a new broker", controllerId, controllerId2)
     TestUtils.waitUntilTrue(() => {
       val metadataResponse2 = sendMetadataRequest(MetadataRequest.allTopics(), 1)
-      controllerServer2.apis.brokerId == metadataResponse2.controller.id
+      metadataResponse2.controller() != null && controllerServer2.apis.brokerId == metadataResponse2.controller.id
```

Comment:
@ijuma 
In this test case I fail over the controller broker right before sending a MetadataRequest with the code below:
```scala
controllerServer.shutdown()
controllerServer.startup()
```

This also happens to be the broker which receives the MetadataRequest in the `waitUntilTrue` block. This request is being sent before the MetadataCache can be updated with the current controllerId. Its a bit of a race condition, which is also why the tests passed locally on my machine. 

The race condition of a MetadataRequest being served before the UpdateMetadataRequest is received has always existed, this just exposed it a bit. It could be solved by requesting the metadata from the controller on initialization, but thats a separate jira and a small edge case. 

The more interesting part is why the test never failed before. This is because the MetadataCache was not reset on restart (`shutdown()` then `startup()`).

In order to fix the broker id problem, I moved the cache initialization into the startup block. Before, the broker would restart and still have the old metadata state until an UpdateMetadataRequest was received. This actually means the information, especially the controllerId was stale and incorrect (still 0 in this case). Now on startup a new MetadataCache is instantiated meaning the controllerId will be None/Null until the first UpdateMetadataRequest is received. 

Using the `shutdown()` and `startup()` commands is something that should really only occur in a test environment, so the stale cache problem shouldn't have impacted real deployments.